include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -3,428 +3,573 @@
 #else
-		return glm::findLSB(val);
+		Vector output;
+		using traits = hlsl::vector_traits<Vector>;
+		for (uint32_t i = 0; i < traits::Dimension; ++i)
+			output[i] = bitReverse_helper<traits::scalar_type>::__call(vec[i]);
+		return output;
 #endif
 	}
 };
 
-#ifndef __HLSL_VERSION
-
-template<typename EnumType>
-requires std::is_enum_v<EnumType>
-struct find_lsb_helper<EnumType>
+template<typename Vector>
+NBL_PARTIAL_REQ_TOP(is_vector_v<Vector>)
+struct min_helper<Vector NBL_PARTIAL_REQ_BOT(is_vector_v<Vector>) >
 {
-	static int32_t findLSB(NBL_CONST_REF_ARG(EnumType) val)
+	static Vector __call(NBL_CONST_REF_ARG(Vector) a, NBL_CONST_REF_ARG(Vector) b)
 	{
-		using underlying_t = std::underlying_type_t<EnumType>;
-		return find_lsb_helper<underlying_t>::findLSB(static_cast<underlying_t>(val));
-	}
-};
+		using traits = hlsl::vector_traits<Vector>;
+		array_get<Vector, typename traits::scalar_type> getter;
+		array_set<Vector, typename traits::scalar_type> setter;
 
-#endif
+		Vector output;
+		for (uint32_t i = 0; i < traits::Dimension; ++i)
+			setter(output, i, min_helper<typename traits::scalar_type>::__call(getter(a, i), getter(b, i)));
 
-template<typename Integer>
-struct find_msb_return_type
-{
-	using type = int32_t;
+		return output;
+	}
 };
-template<typename Integer, int N>
-struct find_msb_return_type<vector<Integer, N> >
+template<typename Vector>
+NBL_PARTIAL_REQ_TOP(concepts::Vectorial<Vector>)
+struct max_helper<Vector NBL_PARTIAL_REQ_BOT(concepts::Vectorial<Vector>) >
 {
-	using type = vector<int32_t, N>;
-};
-template<typename Integer>
-using find_lsb_return_type = find_msb_return_type<Integer>;
+	static Vector __call(NBL_CONST_REF_ARG(Vector) a, NBL_CONST_REF_ARG(Vector) b)
+	{
+		using traits = hlsl::vector_traits<Vector>;
+		array_get<Vector, typename traits::scalar_type> getter;
+		array_set<Vector, typename traits::scalar_type> setter;
 
-template<typename T NBL_STRUCT_CONSTRAINABLE>
-struct bitReverse_helper;
+		Vector output;
+		for (uint32_t i = 0; i < traits::Dimension; ++i)
+			setter(output, i, max_helper<typename traits::scalar_type>::__call(getter(a, i), getter(b, i)));
+
+		return output;
+	}
+};
 
 template<typename Integer>
-NBL_PARTIAL_REQ_TOP(hlsl::is_integral_v<Integer> && hlsl::is_scalar_v<Integer>)
-struct bitReverse_helper<Integer NBL_PARTIAL_REQ_BOT(hlsl::is_integral_v<Integer>&& hlsl::is_scalar_v<Integer>) >
+NBL_PARTIAL_REQ_TOP(concepts::IntegralScalar<Integer>)
+struct bitCount_helper<Integer NBL_PARTIAL_REQ_BOT(concepts::IntegralScalar<Integer>) >
 {
-	static inline Integer __call(NBL_CONST_REF_ARG(Integer) val)
+	using return_t = int32_t;
+	static return_t __call(NBL_CONST_REF_ARG(Integer) val)
 	{
 #ifdef __HLSL_VERSION
-		return spirv::bitReverse(val);
+		if (sizeof(Integer) == 8u)
+		{
+			uint32_t lowBits = uint32_t(val);
+			uint32_t highBits = uint32_t(uint64_t(val) >> 32u);
+
+			return countbits(lowBits) + countbits(highBits);
+		}
+
+		return spirv::bitCount(val);
 #else
```

Comment:
this is kindof off.

Similar to https://github.com/Devsh-Graphics-Programming/Nabla/pull/821/files#r1928536974, the SPIR-V and C++ impl should take care of uint16_t and uint32_t, then emulationh (on hLSL side) should take care of uint64_t.

There's no case for a common specialization here.

Btw don't use `std::bitset`, use c++ `std::popcount` instead,